esp-hal/src/sync.rs

```diff
@@ -106,23 +106,65 @@ mod multicore {
     }
 }
 
-pub(crate) struct Lock {
+cfg_if::cfg_if! {
+    if #[cfg(riscv)] {
+        // The restore state is a u8 that is casted from a bool, so it has a value of
+        // 0x00 or 0x01 before we add the reentry flag to it.
+        pub const REENTRY_FLAG: u8 = 1 << 7;
+    } else if #[cfg(xtensa)] {
+        // PS has 15 useful bits. Bits 12..16 and 19..32 are unused, so we can use bit
+        // #31 as our reentry flag.
+        // We can assume the reserved bit is 0 otherwise rsil - wsr pairings would be
+        // undefined behavior: Quoting the ISA summary, table 64:
+        // Writing a non-zero value to these fields results in undefined processor behavior.
+        pub const REENTRY_FLAG: u32 = 1 << 31;
+    }
+}
+
+/// A lock that can be used to protect shared resources.
+pub struct Lock {
```

Comment:
At this point, might as well combine `Lock` and `Locked` into one type. Anything currently using `Lock` can do `Locked<()>`.

Reply:
I want to explore if we can turn Lock into `MultiCoreCriticalSectionRawMutex` (lovely name!) for embassy-sync, and those raw mutexes don't have data attached to them.

Reply:
It can certainly be done but only with the re-entrant version. `EspRawMutex` would do just fine as a name tbh. (Could probably typedef it to `CriticalSectionRawMutex` on the single core chips :D)

Reply:
> (Could probably typedef it to CriticalSectionRawMutex on the single core chips :D)

That would be a circular definition of what a critical section is :)

Reply:
Ahh I see, I've just better understood what you have in mind